Add new Style/NonLocalControlFlowInTernary cop

Drenmi opened the pull request:
This cop checks for non-local control flow, i.e. `return`, `break`, `next`, `raise`, etc. used in ternary expressions.

-----------------

Before submitting the PR make sure the following are checked:

* [X] Wrote [good commit messages][1].
* [ ] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [X] Feature branch is up-to-date with `master` (if not - rebase it).
* [X] Squashed related commits together.
* [X] Added tests.
* [X] Added an entry to the [Changelog](https://github.com/rubocop-hq/rubocop/blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop-hq/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format).
* [X] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [X] Run `bundle exec rake default`. It executes all tests and RuboCop for itself, and generates the documentation.

[1]: https://chris.beams.io/posts/git-commit/

bbatsov commented:
This will also need an entry in the style guide.

rrosenblum commented:
What if control flow exists on both sides of the ternary? Should it still be flagged? `foo? ? next 1 : next 2` Or `foo? ? break : next`? Or `foo? ? raise 'Error 1' : raise 'Error 2'`?

Drenmi commented:
> This will also need an entry in the style guide.

👍 Will add when I have some time.

> What if control flow exists on both sides of the ternary? Should it still be flagged?

Yes. I should add a test case for that. 🙂

bbatsov commented:
@Drenmi ping :-)

Drenmi commented:
@bbatsov I have thought about it, and I think it's worth generalizing beyond just ternary expressions.

I opened a PR to The Ruby Style Guide [here](https://github.com/rubocop-hq/ruby-style-guide/pull/785).

koic commented:
It may not be a problem when limited to ternary operator. 
On the other hand, I commented on the Style Guide about cases that I found difficult to generalize.
https://github.com/rubocop-hq/ruby-style-guide/pull/785#issuecomment-569915592

Drenmi commented:
@koic @bbatsov I have rebased this and added the spec that @rrosenblum suggested. I think we can merge this. I suspect the Style Guide PR will take some time to iron out. 🙂

koic commented on config/default.yml:
```diff
@@ -3284,6 +3284,11 @@ Style/NilComparison:
+Style/NonLocalControlFlowInTernary:
+  Description: 'checks for non-local control flow in ternary expressions.'
+  Enabled: true
+  VersionAdded: '0.76'
```
It looks like the version needs to be updated.

```suggestion
  VersionAdded: '0.81'
```

koic commented on lib/rubocop/cop/style/non_local_control_flow_in_ternary.rb:
```diff
@@ -0,0 +1,46 @@
+# frozen_string_literal: true
+
+# TODO: when finished, run `rake generate_cops_documentation` to update the docs
```
Can you remove the auto-generated line?

bbatsov commented:
I'm closing this PR due to no recent activity. Feel free to re-open it if you ever come back to it.
